File: 25_1/s_15/exo9.java
public class exo9 {

    /*
      Problem 1:
      Count number of finite functions:
      - {0,1}^n -> {0,1}
      - {0,1}^n -> {-1,0,1}
      - {0,1}^n -> {0,1}^m

      Problem 2:
     AND, OR, NOT using NAND

      Problem 3:
      idea why any Boolean function is computable
    */

    //Problem 1
    static void problem1(int n, int m) {
        int inputs = 1 << n;
        double f01 = Math.pow(2, inputs);
        double f3 = Math.pow(3, inputs);
        double f01m = Math.pow(2, m * inputs);
        System.out.println("Problem 1:");
        System.out.println("Functions {0,1}^" + n + " -> {0,1}: " + (long) f01);
        System.out.println("Functions {0,1}^" + n + " -> {-1,0,1}: " + (long) f3);
        System.out.println("Functions {0,1}^" + n + " -> {0,1}^" + m + ": " + (long) f01m);
        System.out.println();
    }

    //Problem 2
    static int NAND(int a, int b) {
        return (a == 1 && b == 1) ? 0 : 1;}
    static int NOT(int a) {
        return NAND(a, a);}
    static int AND(int a, int b) {
        return NOT(NAND(a, b));
    }
    static int OR(int a, int b) {
        return NAND(NOT(a), NOT(b));
    }
    static void problem2() {
        System.out.println("Problem 2:");
        for (int a = 0; a <= 1; a++) {
            for (int b = 0; b <= 1; b++) {
                System.out.println(
                    "A=" + a + " B=" + b +
                    " AND=" + AND(a,b) +
                    " OR=" + OR(a,b) +
                    " NOT(A)=" + NOT(a));
            }
        }
        System.out.println();
    }

    //Problem 3
    static void problem3(int n) {
        int inputs = 1 << n;
        long functions = (long) Math.pow(2, inputs);
        System.out.println("Problem 3:");
        System.out.println("Any Boolean function {0,1}^" + n + "-> {0,1}");
        System.out.println("Total possible functions: " + functions);
        System.out.println("Each function can be built using delta_x circuits");
        System.out.println("Circuit size: O(n * 2^n)");
        System.out.println();
    }

    public static void main(String[] args) {
        int n = 3;
        int m = 2;
        problem1(n, m);
        problem2();
        problem3(n);
    }
}

File: 25_1/s_7/Exo9.py
import math

# --- Problem 2: Logic Gates from NAND ---
# The task asks to generate AND, OR, NOT using only NAND.
# Here is the implementation.

def nand_gate(a, b):
    # Base NAND operation: 0 only if both inputs are 1
    if a == 1 and b == 1:
        return 0
    return 1

def not_gate(a):
    # NOT A is equivalent to A NAND A
    return nand_gate(a, a)

def and_gate(a, b):
    # AND(A, B) is equivalent to NOT(NAND(A, B))
    # Which means: (A NAND B) NAND (A NAND B)
    n1 = nand_gate(a, b)
    return nand_gate(n1, n1)

def or_gate(a, b):
    # OR(A, B) is equivalent to (NOT A) NAND (NOT B)
    # Which means: (A NAND A) NAND (B NAND B)
    na = nand_gate(a, a)
    nb = nand_gate(b, b)
    return nand_gate(na, nb)

# Simple test for the gates
def test_logic():
    print("--- Problem 2 Checks ---")
    print(f"NOT 0 = {not_gate(0)}") # Should be 1
    print(f"NOT 1 = {not_gate(1)}") # Should be 0
    print(f"AND 1,0 = {and_gate(1, 0)}") # Should be 0
    print(f"OR  0,0 = {or_gate(0, 0)}")  # Should be 0
    print(f"OR  1,0 = {or_gate(1, 0)}")  # Should be 1
    print("Logic checks passed.\n")


# --- Problem 1: Finite Function Possibilities ---
# The code below verifies the formulas for the number of possible functions.
# n = inputs, output set size changes.

def check_possibilities(n, m=1):
    print("--- Problem 1 Calculations ---")
    
    # Total distinct inputs for n bits is 2^n
    total_inputs = 2 ** n
    print(f"For n={n}, total input combinations (leaves) = {total_inputs}")

    # Case 1: Output is {0, 1} (size 2)
    # Formula: 2^(2^n)
    res1 = 2 ** total_inputs
    print(f"1. Output {{0,1}}: 2^(2^{n}) = {res1}")

    # Case 2: Output is {-1, 0, 1} (size 3)
    # Formula: 3^(2^n)
    res2 = 3 ** total_inputs
    print(f"2. Output {{-1,0,1}}: 3^(2^{n}) = {res2}")

    # Case 3: Output is {0, 1}^m (size 2^m)
    # Formula: (2^m)^(2^n) which simplifies to 2^(m * 2^n)
    res3 = (2 ** m) ** total_inputs
    print(f"3. Output {{0,1}}^{m}: 2^({m} * 2^{n}) = {res3}")
    print("\n")

'''
--- Problem 3 Explanation (Theory) ---

We need to realize function F: {0,1}^n -> {0,1}.
Using the Shannon expansion (or Disjunctive Normal Form):

F(x) can be written as an OR of ANDs (minterms).
1. For each input combination (v) where F(v) = 1, we create a circuit that detects v.
2. Detecting a specific input v requires checking n bits (using ANDs and NOTs).
   This part takes O(n) size.
3. In the worst case, we have 2^n such combinations (if the output is always 1).
4. Finally, we OR all these results together.

Total size = (Number of combinations) * (Size to check one combination)
Total size = 2^n * O(n) = O(n * 2^n).
'''

if __name__ == "__main__":
    # Running checks for small n to keep numbers readable
    check_possibilities(n=2, m=2) 
    test_logic()
